### PR TITLE
Fix CPU test drift on transformers 5.5.0 and the MLX bitsandbytes stub

### PR DESCRIPTION
**Pull request by danielhanchen**
## Summary

Four CPU tests fail on `unsloth_zoo` main under two upstream conditions that now co-occur in CI:

1. The MLX-on-torch simulation (activated by the `test_mlx_*` suite) installs a permissive bitsandbytes stub into `sys.modules` for the rest of the session.
2. transformers reached 5.5.0.

None of these are production bugs; they are test-harness and upstream-drift issues. This PR fixes all four so the CPU matrix and the cross-repo full-suite integration job go green.

## Changes

- **stubs/bitsandbytes_stub.py**: `_Noop.__init__` now accepts arbitrary args, including the `(name, bases, namespace)` triple Python passes when a stubbed class is used as a base. `vllm_utils` defines `class Linear4bit(bitsandbytes.nn.modules.Linear4bit)` at import time; against the stub that construction previously raised, taking down 17 `test_vllm_to_hf_conversion` tests. Also exposes an `IS_UNSLOTH_STUB` marker.
- **tests/test_temporary_patches_exhaustive.py**: the bitsandbytes signature-pin tests now skip when bitsandbytes resolves to the stub (via the new marker). Pinning the real upstream `Linear4bit.__init__` signature is meaningless against a stub, so this avoids a false DRIFT.
- **mlx/utils.py**: the baseline loss `labels=None` fast path divides by raw `ntoks` again, to stay byte-for-byte equivalent to `mlx_lm.tuner.trainer.default_loss`. The `_safe_token_denominator` wrapper (fp32 cast plus maximum) stays on the labels-aware path, where labels can be all `-100` and `ntoks` can legitimately be 0.
- **tests/test_mlx_save_export_regressions.py**: the `_download` shim now accepts `allow_patterns`, which `from_pretrained`'s config download forwards upstream. Without it the shim raised, the adapter base-remap path fell through to a direct load, and the test surfaced an unrelated `lazy` kwarg error on transformers 5.5.0.

## Failing tests fixed (20)

- `test_vllm_to_hf_conversion.py` (17 nodes)
- `test_temporary_patches_exhaustive.py::test_bitsandbytes_linear4bit_init_signature`
- `test_mlx_save_export_regressions.py::test_adapter_bnb_base_remap_defaults_to_mlx_4bit_quantization`
- `test_mlx_baseline_loss_parity.py::test_no_safe_token_denominator_in_fast_path`

## Validation

On transformers 5.5.0, the four groups run together in the CI-reproducing order (an mlx module first, which leaks the bnb stub, then the vllm and drift tests):

```
pytest tests/test_mlx_baseline_loss_parity.py \
       tests/test_mlx_save_export_regressions.py \
       tests/test_vllm_to_hf_conversion.py \
       tests/test_temporary_patches_exhaustive.py::test_bitsandbytes_linear4bit_init_signature
-> 100 passed
```

The drift test passes against real bitsandbytes and skips against the stub. No production code path changes behavior: the stub edit only affects the Apple Silicon / MLX stub, and the loss change restores the documented mlx-lm parity guarded by `test_no_safe_token_denominator_in_fast_path`.

### CODE REVIEW

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py [hunk 1]
```diff
@@ -844,7 +844,9 @@ def loss_fn(model, batch, lengths, labels=None):
             mask = mx.logical_and(steps >= lengths[:, 0:1], steps < lengths[:, 1:])
             ce = nn.losses.cross_entropy(logits, targets) * mask
             ntoks = mask.sum()
-            ce = ce.astype(mx.float32).sum() / _safe_token_denominator(ntoks)
+            # Raw ntoks (no safe denominator) to match mlx_lm default_loss
+            # byte-for-byte; the safe wrapper stays on the labels-aware path.
+            ce = ce.astype(mx.float32).sum() / ntoks
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py:849]: Confirmed reachable with use_cce=False plus distributed eval (world_size >= 2): a tail global batch can leave a rank with only distributed_pad_mode="empty" padding rows (lengths [0, 0], ntoks 0), so the raw ntoks fast path returns NaN and NaN * 0 poisons the all-sum. Fixed at the aggregation site with an mx.where zero-token guard so the loss_fn parity path stays byte-for-byte. Thanks for the catch.

**Comment by danielhanchen** [comment]
@codex review

File: unsloth_zoo/mlx/utils.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/mlx/utils.py:849]: Correct and reachable. #886 already merged with the raw ntoks parity change, so the guard lands in the follow-up #888: mask the zero-token contribution at the eval aggregation site with mx.where(ntoks > 0, loss * ntoks, 0) so the loss_fn parity stays byte-for-byte while distributed eval no longer divides an empty batch by zero.
